ocaml/xapi/repository.ml

```diff
@@ -177,69 +220,99 @@ let sync ~__context ~self ~token ~token_id =
+        let config_repo yum_conf =
+          match yum_conf with
+          | Some (token_path, yum_plugin) ->
+              (* Configure proxy and token *)
+              let token_param =
+                match token_path with
+                | "" ->
+                    ""
+                | p ->
+                    Printf.sprintf "--setopt=%s.%s=%s" repo_name yum_plugin
+                      (Uri.make ~scheme:"file" ~path:p () |> Uri.to_string)
+              in
+              let proxy_params =
+                let proxy_url_param, proxy_username_param, proxy_password_param
+                    =
+                  get_proxy_params ~__context repo_name
+                in
+                [proxy_url_param; proxy_username_param; proxy_password_param]
+              in
+              let config_repo' params =
+                let Pkg_mgr.{cmd; params} =
+                  "--save" :: params |> fun config ->
+                  Pkgs.config_repo ~repo_name ~config
+                in
+                ignore
+                  (Helpers.call_script ~log_output:Helpers.On_failure cmd params)
+              in
+              config_repo' (token_param :: proxy_params)
           | None ->
-              ""
-        in
-        let proxy_url_param, proxy_username_param, proxy_password_param =
-          get_proxy_params ~__context repo_name
+              ()
         in
-        let Pkg_mgr.{cmd; params} =
-          [
-            "--save"
-          ; proxy_url_param
-          ; proxy_username_param
-          ; proxy_password_param
-          ; token_param
-          ]
-          |> fun config -> Pkgs.config_repo ~repo_name ~config
-        in
-        ignore (Helpers.call_script ~log_output:Helpers.On_failure cmd params) ;
 
-        (* Import YUM repository GPG key to check metadata in reposync *)
-        let Pkg_mgr.{cmd; params} = Pkgs.make_cache ~repo_name in
-        ignore (Helpers.call_script cmd params) ;
+        let make_cache () =
+          (* Import YUM repository GPG key to check metadata in reposync *)
+          let Pkg_mgr.{cmd; params} = Pkgs.make_cache ~repo_name in
+          ignore (Helpers.call_script cmd params)
+        in
 
         (* Sync with remote repository *)
-        let Pkg_mgr.{cmd; params} = Pkgs.sync_repo ~repo_name in
-        Unixext.mkdir_rec !Xapi_globs.local_pool_repo_dir 0o700 ;
-        clean_yum_cache repo_name ;
-        ignore (Helpers.call_script cmd params)
+        let sync_repo () =
+          let Pkg_mgr.{cmd; params} = Pkgs.sync_repo ~repo_name in
+          Unixext.mkdir_rec !Xapi_globs.local_pool_repo_dir 0o700 ;
+          clean_yum_cache repo_name ;
+          ignore (Helpers.call_script cmd params)
+        in
+
+        with_sync_client_auth client_auth @@ fun yum_conf ->
+        with_sync_server_auth server_auth @@ fun binary_url' ->
+        write_initial_yum_config
+          ~binary_url:(Option.value binary_url' ~default:binary_url) ;
+        config_repo yum_conf ;
```

Comment:
The `config_repo` is working as accept a client auth tuple and add proxy under the hood.
Instead, the `yum_conf` can be just a prepared sync client params, e.g. with name `client_auth_params`. And proxy can be another sync client params prepared outside of the `config_repo`, e.g. with name `client_proxy_params`. Then the `config_repo` is just to accept these params to run the command  with all the parameters.